### PR TITLE
Thy09/use jupyter inspect to get signature of dynamic functions

### PR DESCRIPTION
**Pull request by thy09**
We try to get the signature help from Jupyter kernel for a function which is dynamically created in Jupyter notebook,  by calling `activeNotebook.inspect` like what `provideJupyterCompletionItems` does.

In the code, when handling the hover request/signature help request, if the following conditions are satisfied, we will show the inspect result from Jupyter Kernel:
1. The Jupyter kernel provide a callable signature;
2. The language server doesn't return a callable signature, or the signature of callable is (*args, **kwargs);

Result:
In most cases, the hover and the signature help work as before.
For the case the the function is dynamically created like the following:
```
def create_function_with_parameters(f, sig, name, doc):
    f = types.FunctionType(
        code=f.__code__,
        globals=f.__globals__,
        name=name,
        argdefs=f.__defaults__,
        closure=f.__closure__,
    )
    f.__signature__ = sig
    f.__doc__ = doc
    return f
signature = Signature([Parameter(name='str_param', kind=Parameter.KEYWORD_ONLY, default='a', annotation=str)], return_annotation=int)
g = create_function_with_parameters(f, sig=signature, name='g', doc="Some doc")
```

**Before**:
Signature help:
image
Hover:
image

**After**:
Signature help:
image
Hover:
image

-   [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [ ] Title summarizes what is changing.
-   [ ] Has a [news entry](https://github.com/Microsoft/vscode-python/tree/master/news) file (remember to thank yourself!).
-   [ ] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for enhancements.
-   [ ] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-python/blob/master/.github/test_plan.md) is updated as appropriate.
-   [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-python/blob/master/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).
-   [ ] The wiki is updated with any design decisions/details.

### CODE REVIEW

File: src/client/datascience/interactive-common/intellisense/intellisenseProvider.ts [hunk 1]
```diff
@@ -585,20 +698,55 @@ export class IntellisenseProvider implements IInteractiveWindowListener {
...
+                request.cellId,
+                cancelSource.token
+            );
+
+            // Capture telemetry for each of the two providers.
+            // Telemetry will be used to improve how we handle intellisense to improve response times for code completion.
+            // NOTE: If this code is around after a few months, telemetry isn't used, or we don't need it anymore.
+            // I.e. delete this code.
+            sendTelemetryWhenDone(Telemetry.CompletionTimeFromLS, lsSignatureHelp);
+            sendTelemetryWhenDone(Telemetry.CompletionTimeFromJupyter, jupyterSignatureHelp);
```
Comment (wangchao1230) [inline src/client/datascience/interactive-common/intellisense/intellisenseProvider.ts]: seems these are for completion not signature help?

File: src/client/datascience/jupyter/jupyterNotebook.ts [hunk 2]
```diff
@@ -354,7 +354,7 @@ export class JupyterNotebookBase implements INotebook {
         return deferred.promise;
     }
 
-    public inspect(code: string, cancelToken?: CancellationToken): Promise<JSONObject> {
+    public inspect(code: string, offsetInCode = 0, cancelToken?: CancellationToken): Promise<JSONObject> {
```
Comment (wangchao1230) [inline src/client/datascience/jupyter/jupyterNotebook.ts:357]: shall we put this param to the end, so no breaking to exsiting code?

Comment (thy09) [reply src/client/datascience/jupyter/jupyterNotebook.ts:357]: In this case, there are 2 concerns:
1. The interface will be different to `getCompletion`, which also calls jupyter kernel.
2. It's hard to provide enough information for hover request, for example: the code is "module_func()", while the cursor is at '', if we simply truncate the string, we can only get "module_f", which cannot get anything. Or we need to infer which word is required, which is complicated and break the design of jupyter inspect API.

File: src/client/datascience/interactive-common/intellisense/intellisenseProvider.ts (hunk 1 above)
Comment (thy09) [reply src/client/datascience/interactive-common/intellisense/intellisenseProvider.ts]: Let me have a check

Comment (thy09) [reply src/client/datascience/interactive-common/intellisense/intellisenseProvider.ts]: Removed

**Pull request merged**